8918: update server 2 server auth docs (#9056)
CDCgov/prime-reportstream: prime-router/docs/features/0002-authentication.md

# ReportStream authentication

## Summary

This document covers how authentication is handled by ReportStream.  Authentication is broken down into two implementations:

- Authentication for users who can log into the system via Okta
- Authentication based on SMART on FHIR OAuth guide

The entry point for this authentication is [here](https://github.com/CDCgov/prime-reportstream/blob/403347e889c2fb614659ee626e8c0fcdc2cf6c93/prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt#L198);
the code will attempt to first authenticate via Okta and then the SMART on FHIR OAuth implementation

### via Okta

Users with credentials via Okta (typically those who can log in to the report stream UI) login via Okta and a short-lived token is returned and users include that in requests to the API.  ReportStream uses the Okta JWT library to authenticate
and [decode](https://github.com/CDCgov/prime-reportstream/blob/403347e889c2fb614659ee626e8c0fcdc2cf6c93/prime-router/src/main/kotlin/tokens/OktaAuthentication.kt#L69) the JWT.

Most of this happens without a user necessarily being aware of the token as it's handled by the browser.

### via SMART on FHIR oauth implementation

> This implementation follows the guide linked below and it essentially an implementation of an Oauth flow and the image 

Reviewer: and it `is` essentially

> This implementation follows the guide linked below and it essentially an implementation of an Oauth flow and the image 

Author: ```suggestion
This implementation follows the guide linked below and it is essentially an implementation of an Oauth flow and the image 
```

below shows the flow.  At a high level:

> 

Reviewer: Maybe add step one as being "client generates a key pair" and have that link to our instructions on where we talk about how that is done?

- the client uploads the public key for the key pair they generate

> - they sign a JWT (aka a JWS) with the that private key

Reviewer: `with their private key`

> - they sign a JWT (aka a JWS) with the that private key

Author: ```suggestion
- they sign a JWT (aka a JWS) with their private key
```
